features/cellular/TARGET_GENERIC_MODEM/generic_modem_driver/GenericCellularInterface.h

```diff
@@ -0,0 +1,279 @@
+     *  the lookup table then the driver tries to resort to default APN settings.
+     *
+     *  Preferred method is to setup APN using 'set_credentials()' API.
+
+     *  @return         0 on success, negative error code on failure
+     */
+    virtual nsapi_error_t connect();
+
+    /** Attempt to disconnect from the network
+     *
+     *  Brings down the network interface. Shuts down the PPP interface
+     *  of the underlying network stack. Does not bring down the Radio network
+     *
+     *  @return         0 on success, negative error code on failure
+     */
+    virtual nsapi_error_t disconnect();
+
+    /** Adds or removes a SIM facility lock
+     *
+     * Can be used to enable or disable SIM pin check at device startup.
+     * This API sets up flags for the driver which would either enable or disable
+     * SIM pin checking depending upon the user provided argument while establishing
+     * connection. It doesn't do anything immediately other than setting up flags.
+     *
+     * @param set        can be set to true if the SIM pin check is supposed to be enabled
+     *                   and vice versa.
+     */
+    void set_sim_pin_check(bool set);
+
+    /** Change the pin for the SIM card
+     *
+     * Provide the new pin for your SIM card with this API. Old pin code will be assumed to
+     * be set using set_SIM_pin() API. This API have no immediate effect. While establishing
+     * connection, driver will change the SIM pin for the next boot.
+     *
+     * @param new_pin     new pin to be used in string format
+     */
+    void set_new_sim_pin(const char *new_pin);
+
+    /** Check if the connection is currently established or not
+     *
+     * @return true/false   If the cellular module have successfully acquired a carrier and is
+     *                      connected to an external packet data network using PPP, isConnected()
+     *                      API returns true and false otherwise.
+     */
+    virtual bool is_connected();
+
+    /** Get the local IP address
+     *
+     *  @return         Null-terminated representation of the local IP address
+     *                  or null if no IP address has been recieved
+     */
+    virtual const char *get_ip_address();
+
+    /** Get the local network mask
+     *
+     *  @return         Null-terminated representation of the local network mask
+     *                  or null if no network mask has been recieved
+     */
+    virtual const char *get_netmask();
+
+    /** Get the local gateways
+     *
+     *  @return         Null-terminated representation of the local gateway
+     *                  or null if no network mask has been recieved
+     */
+    virtual const char *get_gateway();
+
+    /** Get notified if the connection gets lost
+     *
+     *  @param cb         user defined callback
+     */
+    void connection_status_cb(Callback<void(nsapi_error_t)> cb);
```

Comment:
This will be used and extended by NBIoT team. I was thinking actually to be part of CellularBase class, but then I stopped short of it as that would ignite another flurry of discussions.